doc/dotlock.man
.\" -*-nroff-*-
.\"
.\"
.\"     Copyright (C) 1996-8 Michael R. Elkins <me@cs.hmc.edu>
.\"	Copyright (C) 1998-9 Thomas Roessler <roessler@does-not-exist.org>
.\"
.\"     This program is free software; you can redistribute it and/or modify
.\"     it under the terms of the GNU General Public License as published by
.\"     the Free Software Foundation; either version 2 of the License, or
.\"     (at your option) any later version.
.\"
.\"     This program is distributed in the hope that it will be useful,
.\"     but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"     GNU General Public License for more details.
.\"
.\"     You should have received a copy of the GNU General Public License
.\"     along with this program; if not, write to the Free Software
.\"     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
.\"
.TH dotlock 1 "AUGUST 1999" Unix "User Manuals"
.SH NAME
mutt_dotlock \- Lock mail spool files.
.SH SYNOPSIS
.PP
.B mutt_dotlock
[\-t|\-f|\-u|\-d] [\-p] [\-r \fIretries\fP] \fIfile\fP
.SH DESCRIPTION
.PP
.B mutt_dotlock
implements the traditional mail spool file locking method:
To lock \fIfile\fP, a file named \fIfile\fP.lock is
created. The program operates with group mail privileges
if necessary.
.SH OPTIONS
.PP
.IP "-t"
Just try.
.B mutt_dotlock
won't actually lock a file, but inform the invoking
process if it's at all possible to lock \fIfile\fP.
.IP "-f"
Force the lock.  If another process holds a lock on
\fIfile\fP longer than a certain amount of time,
.B mutt_dotlock
will break that lock by removing the lockfile.
.IP "-u"
Unlock.
.B mutt_dotlock
will remove \fIfile\fP.lock.
.IP "-d"
Delete.
.B mutt_dotlock
will lock \fIfile\fP, remove it if it has length 0, and afterwards
remove \fIfile\fP.lock.
.IP "-p"
Use privileges.  If given this option,
.B mutt_dotlock
will operate with group mail privileges when creating and
deleting lock files.
.IP "-r \fIretries\fP"
This command line option tells
.B mutt_dotlock
to try locking
\fIretries\fP times before giving up or (if invoked with
the
.B -f
command line option) break a lock.  The default value is 5.
.B mutt_dotlock
waits one second between successive locking attempts.
.SH FILES
.PP
.IP "\fIfile\fP.lock"
The lock file
.B mutt_dotlock
generates.
.SH SEE ALSO
.PP
.BR fcntl (2),
.BR flock (2),
.BR lockfile (1),
.BR mutt (1)
.SH DIAGNOSTICS
.PP
.B mutt_dotlock
gives all diagnostics in its return values:
.TP
.B "0 \- DL_EX_OK"
The program was successful.
.TP
.B "1 \- DL_EX_ERROR"
An unspecified error such as bad command line parameters,
lack of system memory and the like has occurred.
.TP
.B "3 \- DL_EX_EXIST"
The
user wants to lock a file which has been locked by
another process already.  If
.B mutt_dotlock
is invoked with the
.B -f
command line option,
.B mutt_dotlock
won't generate this error, but break other processes'
locks.
.TP
.B "4 \- DL_EX_NEED_RPIVS"
This return value only occurs if
.B mutt_dotlock
has been invoked
with the
.B -t
command line option.  It means that
.B mutt_dotlock
will have to use its group mail privileges to lock
\fIfile\fP.
.TP
.B "5 \- DL_EX_IMPOSSIBLE"
This return value only occurs if
.B mutt_dotlock
has been invoked with the
.B -t
command line option.  It means that
.B mutt_dotlock
is unable to lock \fIfile\fP even with group mail
privileges.
.SH NOTES
.PP
.B mutt_dotlock
tries to implement an NFS-safe dotlocking method which was
borrowed from
.B lockfile
(1).
.PP
If the user can't open \fIfile\fP for reading with his
normal privileges,
.B mutt_dotlock
will return the
.B DL_EX_ERROR
exit value to avoid certain attacks against other users'
spool files. The code carefully avoids race conditions
when checking permissions; for details of all this see the
comments in dotlock.c.
.SH HISTORY
.PP
.B mutt_dotlock
is part of the Mutt mail user agent package.  It has been
created to avoid running mutt with group mail privileges.
.SH AUTHOR
Thomas Roessler <roessler@does-not-exist.org>


NEWS
		   Visible changes since Mutt 1.2
		   ==============================


Folder formats and folder access
--------------------------------

- Better mh support: Mutt now supports .mh_sequences files.
  Currently, the "unseen", "flagged", and "replied" sequences are
  used to store mutt flags (the names are configurable using the
  $mh_seq_unseen, $mh_seq_flagged, and $mh_seq_replied configuration
  variables). As a side effect, messages in MH folders are no longer
  rewritten upon status changes.

- The "trashed" flag is supported for maildir folders.  See
  $maildir_trash.

- POP folder support.  You can now access a POP mailbox just like an
  IMAP folder (with obvious restrictions due to the protocol).

- URL syntax for remote folders.  You can pass things like
  pop://account@host and imap://account@host/folder as arguments for
  the -f command line flag.

- STARTTLS support.  If $ssl_starttls is set (the default), mutt
  will attempt to use STARTTLS on servers advertising that
  capability.

- $preconnect.  If set, a shell command to be executed if mutt fails
  to establish a connection to the server.  This is useful for
  setting up secure connections; see the muttrc(5) for details.

- $tunnel.  Use a pipe to a command instead of a raw socket.  See
  muttrc(5) for details.  (Basically, it's another way for setting
  up secure connections.)

- More new IMAP/POP-related variables (see muttrc(5) for details):
  $connect_timeout, $imap_authenticators, $imap_delim_chars,
  $imap_peek, $pop_authenticators, $pop_auth_try_all,
  $pop_checkinterval, $pop_delete, $pop_reconnect, $use_ipv6.

- The following IMAP/POP-related variables are gone:
  $imap_checkinterval, $imap_cramkey, $pop_port.

- There's a new imap-fetch-mail function, which forces a check for
  new messages on an IMAP server.

- The new-mailbox function was renamed to create-mailbox, and is
  bound to C instead of n by default.

Character set support
---------------------

- Mutt now uses the iconv interface for character set conversions.
  This means that you need either a very modern libc, or Bruno
  Haible's libiconv, which is available from
  <http://www.gnu.org/software/libiconv/>.

- With sufficiently recent versions of ncurses and slang, mutt works
  properly in utf-8 locales.

- On sufficiently modern systems, the $charset variable's value is
  automatically derived from the locale you use.  (Note, however,
  that manually setting it to a value which is compatible with your
  locale doesn't do any harm.)

- $send_charset is a colon-separated list of character sets now,
  defaulting to us-ascii:iso-8859-1:utf-8.

- charset-hook defines aliases for character sets encountered in
  messages (say, someone tags his messages with latin15 when he
  means iso-8859-15), iconv-hook defines local names for character
  sets (for systems which don't know about MIME names; see
  contrib/iconv for sample configuration snippets).

- The change-charset function is gone.  Use edit-type (C-e on the
  compose menu) instead.

- The recode-attachment function is gone.

Other changes
-------------

- There's a new variable $compose_format for the compose screen's
  status line.  You can now include the message's approximate
  on-the-wire size.

- The attachment menu knows about collapsing now: Using
  collapse-parts (bound to "v" by default), you can collapse and
  uncollapse parts of the attachment tree.  This function is also
  available from the pager when invoked from the attachment tree.

  Normally, the recvattach menu will start uncollapsed.  However,
  with the new $digest_collapse option (which is set by default),
  the individual messages contained in digests will be displayed
  collapsed.  (That is, there's one line per message.)

- Using $display_filter, you can specify a command which filters
  messages before they are displayed.

- Using message-hook, you can execute mutt configuration commands
  before a message is displayed (or formatted before replying).

- If you don't want that mutt moves flagged messages to your mbox,
  set $keep_flagged.

- Setting the $pgp_ignore_subkeys variable will cause mutt to ignore
  OpenPGP.  This option is set by default, and it's suggested that
  you leave it.

- $pgp_sign_micalg has gone.  Mutt now automatically determines what
  MIC algorithm was used for a particular signature.

- If $pgp_good_sign is set, then a PGP signature is only considered
  verified if the output from $pgp_verify_command matches this
  regular expression.  It's suggested that you set this variable to
  the typical text message output by PGP (or GPG, or whatever)
  produces when it encounters a good signature.

- There's a new function, check-traditional-pgp, which is bound to
  esc-P by default.  It'll check whether a text parts of a message
  contain PGP encrypted or signed material, and possibly adjust
  content types.

- $print_split.  If this option is set, $print_command run
  separately for each message you print.  Useful with enscript(1)'s
  mail printing mode.

- $sig_on_top.  Include the signature before any quoted or forwarded
  text.  WARNING: use of this option may provoke flames.

- $text_flowed.  When set, mutt will generate text/plain attachments
  with the format=flowed parameter.  In order to properly produce
  such messages, you'll need an appropriate editor mode.  Note that
  the $indent_string option is ignored with flowed text.

- $to_chars has grown:  Mailing list messages are now tagged with an
  L in the index. If you want the old behaviour back, add this to
  your .muttrc: set to_chars=" +TCF "

- New emacs-like functions in the line editor: backward-word (M-b),
  capitalize-word (M-c), downcase-word (M-l), upcase-word (M-u),
  forward-word (M-f), kill-eow (M-d), tranpose-chars (unbound).

  transpose-chars is unbound by default because external query
  occupies C-t.  Suggested alternative binding:

	bind	editor	"\e\t"		complete-query
	bind	editor	"\Ct"		transpose-chars


dotlock.c
/*
 * Copyright (C) 1996-2000 Michael R. Elkins <me@mutt.org>
 * Copyright (C) 1998-2001,2007 Thomas Roessler <roessler@does-not-exist.org>
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

/*
 * This module either be compiled into Mutt, or it can be
 * built as a separate program. For building it
 * separately, define the DL_STANDALONE preprocessor
 * macro.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <unistd.h>
#include <dirent.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <sys/utsname.h>
#include <errno.h>
#include <time.h>
#include <fcntl.h>
#include <limits.h>

#ifndef _POSIX_PATH_MAX
#include <limits.h>
#endif

#include "dotlock.h"

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#ifdef DL_STANDALONE
#include "git-ver.h"
#endif

#define MAXLINKS 1024 /* maximum link depth */

#ifdef DL_STANDALONE

# define LONG_STRING 1024
# define MAXLOCKATTEMPT 5

#define strfcpy(DST,SRC,LEN) do { if ((LEN) > 0) { *(DST+(LEN)-1)=0; strncpy(DST,SRC,(LEN)-1); } } while (0)

# ifdef USE_SETGID

#  ifdef HAVE_SETEGID
#   define SETEGID setegid
#  else
#   define SETEGID setgid
#  endif
#  ifndef S_ISLNK
#   define S_ISLNK(x) (((x) & S_IFMT) == S_IFLNK ? 1 : 0)
#  endif

# endif

#else  /* DL_STANDALONE */

# ifdef USE_SETGID
#   error Do not try to compile dotlock as a mutt module when requiring egid switching!
# endif

# include "mutt.h"
# include "mx.h"

#endif /* DL_STANDALONE */

static int DotlockFlags;
static int Retry = MAXLOCKATTEMPT;

#ifdef DL_STANDALONE
struct utsname utsname;
#endif

#ifdef USE_SETGID
static gid_t UserGid;
static gid_t MailGid;
#endif

/* These functions work on the current directory:
 *   dotlock_try dotlock_unlock dotlock_unlink dotlock_lock
 * Invoke dotlock_prepare () before and check their
 * return value.
 */

#ifdef DL_STANDALONE
/*
 * Determine our effective group ID, and drop
 * privileges.
 *
 * Return value:
 *
 *  0 - everything went fine
 * -1 - we couldn't drop privileges.
 *
 */
static int
dotlock_init_privs (void)
{

# ifdef USE_SETGID

  UserGid = getgid ();
  MailGid = getegid ();

  if (SETEGID (UserGid) != 0)
    return -1;

# endif

  return 0;
}
#endif /* DL_STANDALONE */

/*
 * Get privileges
 *
 * This function re-acquires the privileges we may have
 * if the user told us to do so by giving the "-p"
 * command line option.
 *
 * BEGIN_PRIVILEGES () won't return if an error occurs.
 *
 */
static void
BEGIN_PRIVILEGED (void)
{
#ifdef USE_SETGID
  if (DotlockFlags & DL_FL_USEPRIV)
  {
    if (SETEGID (MailGid) != 0)
    {
      exit (DL_EX_ERROR);
    }
  }
#endif
}

/*
 * Drop privileges
 *
 * This function drops the group privileges we may have.
 *
 * END_PRIVILEGED () won't return if an error occurs.
 *
 */
static void
END_PRIVILEGED (void)
{
#ifdef USE_SETGID
  if (DotlockFlags & DL_FL_USEPRIV)
  {
    if (SETEGID (UserGid) != 0)
    {
      exit (DL_EX_ERROR);
    }
  }
#endif
}

#ifdef DL_STANDALONE
/*
 * Usage information.
 *
 * This function doesn't return.
 *
 */
static void
usage (const char *av0)
{
  fprintf (stderr, "dotlock [NeoMutt %s%s (%s)]\n", PACKAGE_VERSION, GitVer, MUTT_VERSION);
  fprintf (stderr, "usage: %s [-t|-f|-u|-d] [-p] [-r <retries>] file\n",
    av0);

  fputs ("\noptions:"
  "\n  -t\t\ttry"
  "\n  -f\t\tforce"
  "\n  -u\t\tunlock"
  "\n  -d\t\tunlink"
  "\n  -p\t\tprivileged"
#ifndef USE_SETGID
  " (ignored)"
#endif
  "\n  -r <retries>\tRetry locking"
  "\n", stderr);

  exit (DL_EX_ERROR);
}
#endif

/*
 * Access checking: Let's avoid to lock other users' mail
 * spool files if we aren't permitted to read them.
 *
 * Some simple-minded access (2) checking isn't sufficient
 * here: The problem is that the user may give us a
 * deeply nested path to a file which has the same name
 * as the file he wants to lock, but different
 * permissions, say, e.g.
 * /tmp/lots/of/subdirs/var/spool/mail/root.
 *
 * He may then try to replace /tmp/lots/of/subdirs by a
 * symbolic link to / after we have invoked access () to
 * check the file's permissions.  The lockfile we'd
 * create or remove would then actually be
 * /var/spool/mail/root.
 *
 * To avoid this attack, we proceed as follows:
 *
 * - First, follow symbolic links a la
 *   dotlock_deference_symlink ().
 *
 * - get the result's dirname.
 *
 * - chdir to this directory.  If you can't, bail out.
 *
 * - try to open the file in question, only using its
 *   basename.  If you can't, bail out.
 *
 * - fstat that file and compare the result to a
 *   subsequent lstat (only using the basename).  If
 *   the comparison fails, bail out.
 *
 * dotlock_prepare () is invoked from main () directly
 * after the command line parsing has been done.
 *
 * Return values:
 *
 * 0 - Everything's fine.  The program's new current
 *     directory is the contains the file to be locked.
 *     The string pointed to by bn contains the name of
 *     the file to be locked.
 *
 * -1 - Something failed. Don't continue.
 *
 * tlr, Jul 15 1998
 */
static int
dotlock_check_stats (struct stat *fsb, struct stat *lsb)
{
  /* S_ISLNK (fsb->st_mode) should actually be impossible,
   * but we may have mixed up the parameters somewhere.
   * play safe.
   */

  if (S_ISLNK (lsb->st_mode) || S_ISLNK (fsb->st_mode))
    return -1;

  if ((lsb->st_dev != fsb->st_dev) ||
     (lsb->st_ino != fsb->st_ino) ||
     (lsb->st_mode != fsb->st_mode) ||
     (lsb->st_nlink != fsb->st_nlink) ||
     (lsb->st_uid != fsb->st_uid) ||
     (lsb->st_gid != fsb->st_gid) ||
     (lsb->st_rdev != fsb->st_rdev) ||
     (lsb->st_size != fsb->st_size))
  {
    /* something's fishy */
    return -1;
  }

  return 0;
}

/*
 * Expand a symbolic link.
 *
 * This function expects newpath to have space for
 * at least _POSIX_PATH_MAX characters.
 *
 */
static void
dotlock_expand_link (char *newpath, const char *path, const char *link)
{
  const char *lb = NULL;
  size_t len;

  /* link is full path */
  if (*link == '/')
  {
    strfcpy (newpath, link, _POSIX_PATH_MAX);
    return;
  }

  if ((lb = strrchr (path, '/')) == NULL)
  {
    /* no path in link */
    strfcpy (newpath, link, _POSIX_PATH_MAX);
    return;
  }

  len = lb - path + 1;
  memcpy (newpath, path, len);
  strfcpy (newpath + len, link, _POSIX_PATH_MAX - len);
}

/*
 * Check if a file can be locked at all.
 *
 * The same comment as for dotlock_lock () applies here.
 *
 */
static int
dotlock_try (void)
{
#ifdef USE_SETGID
  struct stat sb;
#endif

  if (access (".", W_OK) == 0)
    return DL_EX_OK;

#ifdef USE_SETGID
  if (stat (".", &sb) == 0)
  {
    if ((sb.st_mode & S_IWGRP) == S_IWGRP && sb.st_gid == MailGid)
      return DL_EX_NEED_PRIVS;
  }
#endif

  return DL_EX_IMPOSSIBLE;
}

/*
 * Deference a chain of symbolic links
 *
 * The final path is written to d.
 *
 */
static int
dotlock_deference_symlink (char *d, size_t l, const char *path)
{
  struct stat sb;
  char realpath[_POSIX_PATH_MAX];
  const char *pathptr = path;
  int count = 0;

  while (count++ < MAXLINKS)
  {
    if (lstat (pathptr, &sb) == -1)
    {
      return -1;
    }

    if (S_ISLNK (sb.st_mode))
    {
      char linkfile[_POSIX_PATH_MAX];
      char linkpath[_POSIX_PATH_MAX];
      int len;

      if ((len = readlink (pathptr, linkfile, sizeof (linkfile) - 1)) == -1)
      {
	return -1;
      }

      linkfile[len] = '\0';
      dotlock_expand_link (linkpath, pathptr, linkfile);
      strfcpy (realpath, linkpath, sizeof (realpath));
      pathptr = realpath;
    }
    else
      break;
  }

  strfcpy (d, pathptr, l);
  return 0;
}

static int
dotlock_prepare (char *bn, size_t l, const char *f, int _fd)
{
  struct stat fsb, lsb;
  char realpath[_POSIX_PATH_MAX];
  char *basename, *dirname;
  char *p;
  int fd;
  int r;

  if (dotlock_deference_symlink (realpath, sizeof (realpath), f) == -1)
    return -1;

  if ((p = strrchr (realpath, '/')))
  {
    *p = '\0';
    basename = p + 1;
    dirname = realpath;
  }
  else
  {
    basename = realpath;
    dirname = ".";
  }

  if (strlen (basename) + 1 > l)
    return -1;

  strfcpy (bn, basename, l);

  if (chdir (dirname) == -1)
    return -1;

  if (_fd != -1)
    fd = _fd;
  else if ((fd = open (basename, O_RDONLY)) == -1)
    return -1;

  r = fstat (fd, &fsb);

  if (_fd == -1)
    close (fd);

  if (r == -1)
    return -1;

  if (lstat (basename, &lsb) == -1)
    return -1;

  if (dotlock_check_stats (&fsb, &lsb) == -1)
    return -1;

  return 0;
}

#define HARDMAXATTEMPTS 10

/*
 * Dotlock a file.
 *
 * realpath is assumed _not_ to be an absolute path to
 * the file we are about to lock.  Invoke
 * dotlock_prepare () before using this function!
 *
 */
static int
dotlock_lock (const char *realpath)
{
  char lockfile[_POSIX_PATH_MAX + LONG_STRING];
  char nfslockfile[_POSIX_PATH_MAX + LONG_STRING];
  size_t prev_size = 0;
  int fd;
  int count = 0;
  int hard_count = 0;
  struct stat sb;
  time_t t;

  snprintf (nfslockfile, sizeof (nfslockfile), "%s.%s.%d",
       realpath,
#ifdef DL_STANDALONE
       utsname.nodename,
#else
       Hostname,
#endif
       (int) getpid ());
  snprintf (lockfile, sizeof (lockfile), "%s.lock", realpath);

  BEGIN_PRIVILEGED ();

  unlink (nfslockfile);

  while ((fd = open (nfslockfile, O_WRONLY | O_EXCL | O_CREAT, 0)) < 0)
  {
    END_PRIVILEGED ();

    if (errno != EAGAIN)
    {
      return DL_EX_ERROR;
    }

    BEGIN_PRIVILEGED ();
  }

  END_PRIVILEGED ();

  close (fd);

  while (hard_count++ < HARDMAXATTEMPTS)
  {

    BEGIN_PRIVILEGED ();
    link (nfslockfile, lockfile);
    END_PRIVILEGED ();

    if (stat (nfslockfile, &sb) != 0)
    {
      return DL_EX_ERROR;
    }

    if (sb.st_nlink == 2)
      break;

    if (count == 0)
      prev_size = sb.st_size;

    if (prev_size == sb.st_size && ++count > Retry)
    {
      if (DotlockFlags & DL_FL_FORCE)
      {
	BEGIN_PRIVILEGED ();
	unlink (lockfile);
	END_PRIVILEGED ();

	count = 0;
	continue;
      }
      else
      {
	BEGIN_PRIVILEGED ();
	unlink (nfslockfile);
	END_PRIVILEGED ();
	return DL_EX_EXIST;
      }
    }

    prev_size = sb.st_size;

    /* don't trust sleep (3) as it may be interrupted
     * by users sending signals.
     */

    t = time (NULL);
    do {
      sleep (1);
    } while (time (NULL) == t);
  }

  BEGIN_PRIVILEGED ();
  unlink (nfslockfile);
  END_PRIVILEGED ();

  return DL_EX_OK;
}

/*
 * Unlock a file.
 *
 * The same comment as for dotlock_lock () applies here.
 *
 */
static int
dotlock_unlock (const char *realpath)
{
  char lockfile[_POSIX_PATH_MAX + LONG_STRING];
  int i;

  snprintf (lockfile, sizeof (lockfile), "%s.lock",
	   realpath);

  BEGIN_PRIVILEGED ();
  i = unlink (lockfile);
  END_PRIVILEGED ();

  if (i == -1)
    return DL_EX_ERROR;

  return DL_EX_OK;
}

/* remove an empty file */

static int
dotlock_unlink (const char *realpath)
{
  struct stat lsb;
  int i = -1;

  if (dotlock_lock (realpath) != DL_EX_OK)
    return DL_EX_ERROR;

  if ((i = lstat (realpath, &lsb)) == 0 && lsb.st_size == 0)
    unlink (realpath);

  dotlock_unlock (realpath);

  return (i == 0) ?  DL_EX_OK : DL_EX_ERROR;
}

static int dotlock_dispatch (const char *f, int fd)
{
  char realpath[_POSIX_PATH_MAX];

  /* If dotlock_prepare () succeeds [return value == 0],
   * realpath contains the basename of f, and we have
   * successfully changed our working directory to
   * `dirname $f`.  Additionally, f has been opened for
   * reading to verify that the user has at least read
   * permissions on that file.
   *
   * For a more detailed explanation of all this, see the
   * lengthy comment below.
   */

  if (dotlock_prepare (realpath, sizeof (realpath), f, fd) != 0)
    return DL_EX_ERROR;

  /* Actually perform the locking operation. */

  if (DotlockFlags & DL_FL_TRY)
    return dotlock_try ();
  else if (DotlockFlags & DL_FL_UNLOCK)
    return dotlock_unlock (realpath);
  else if (DotlockFlags & DL_FL_UNLINK)
    return dotlock_unlink (realpath);
  else /* lock */
    return dotlock_lock (realpath);
}

#ifndef DL_STANDALONE
/*
 * This function is intended to be invoked from within
 * mutt instead of mx.c's invoke_dotlock ().
 */
int dotlock_invoke (const char *path, int fd, int flags, int retry)
{
  int currdir;
  int r;

  DotlockFlags = flags;

  if ((currdir = open (".", O_RDONLY)) == -1)
    return DL_EX_ERROR;

  if (!(DotlockFlags & DL_FL_RETRY) || retry)
    Retry = MAXLOCKATTEMPT;
  else
    Retry = 0;

  r = dotlock_dispatch (path, fd);

  fchdir (currdir);
  close (currdir);

  return r;
}
#endif  /* !DL_STANDALONE */


#ifdef DL_STANDALONE

#define check_flags(a) if (a & DL_FL_ACTIONS) usage (argv[0])

int main (int argc, char **argv)
{
  int i;
  char *p;

  /* first, drop privileges */

  if (dotlock_init_privs () == -1)
    return DL_EX_ERROR;

  /* determine the system's host name */

  uname (&utsname);
  if ((p = strchr (utsname.nodename, '.')))
    *p = '\0';

  /* parse the command line options. */
  DotlockFlags = 0;

  while ((i = getopt (argc, argv, "dtfupr:")) != EOF)
  {
    switch (i)
    {
      /* actions, mutually exclusive */
      case 't': check_flags (DotlockFlags); DotlockFlags |= DL_FL_TRY; break;
      case 'd': check_flags (DotlockFlags); DotlockFlags |= DL_FL_UNLINK; break;
      case 'u': check_flags (DotlockFlags); DotlockFlags |= DL_FL_UNLOCK; break;

      /* other flags */
      case 'f': DotlockFlags |= DL_FL_FORCE; break;
      case 'p': DotlockFlags |= DL_FL_USEPRIV; break;
      case 'r': DotlockFlags |= DL_FL_RETRY; Retry = atoi (optarg); break;

      default: usage (argv[0]);
    }
  }

  if (optind == argc || Retry < 0)
    usage (argv[0]);

  return dotlock_dispatch (argv[optind], -1);
}
#endif

